### ISSUE TITLE
Value instances and types implementing business logic w/o parsing anything.

### ISSUE THREAD

**Issue by ams-tschoening** [kaitai-io/kaitai_struct#521]
In my KSY-spec I need to support different types of data coming from some IoT-devices and in most cases that data is simply distinguished by some byte: E.g. `0` is an unsigned byte of 8 bits, `1` is a binary-coded decimal of 2 digits etc. But in some cases the devices providing the data are known to contain firmware errors influencing how the data is provided, e.g. it might contain wrongly flipped bits. I have things like manufacturer of a device, its firmware revision and some addressing infos available to distinguish the devices from one another and map broken data to internally used data types to handle them specially later.

The problem is that distinguishing the devices with broken firmwares sometimes is not as easy as only checking the manufacturer and firmware revision, but I need to hard-code a list of serial numbers to check against sometimes and those can easily be some dozen or even hundreds of devices. To improve readability of those checks, I'm trying to implement some concept of types not parsing anything at all, because all necessary infos are available already, and instead only provide special checks using value instances.

The end result simply is something like `get_checker_for_bug_xy.is_affected`, where `is_affected` maps to a lot of hard-coded checks. While this is most likely not what KS has been designed for, it's simply possible in my case and with doing so I would benefit of the same KSY being usable in all supported target languages. I already need to support two languages myself and using parsing instances things seem to work:

    # fmt_oms_apl_data:
    instances:
      dummy:
        pos:  _io.pos
        type: fmt_oms_apl_data_date_g_meh(_root, fmt_oms_dll_device_type::heat_meter_outlet.to_i, fmt_oms_dll_device_type::water_meter.to_i)

`fmt_oms_apl_data_date_g_meh` is the checking type designed to not parse anything at all, but simply provide value instances checking things based on the given data. That translates to the following Java:

        private FmtOmsAplDataDateGMeh dummy;
        public FmtOmsAplDataDateGMeh dummy() {
            if (this.dummy != null)
                return this.dummy;
            long _pos = this._io.pos();
            this._io.seek(_io().pos());
            this.dummy = new FmtOmsAplDataDateGMeh(this._io, _root, FmtRecOms.FmtOmsDllDeviceType.HEAT_METER_OUTLET.id(), FmtRecOms.FmtOmsDllDeviceType.WATER_METER.id());
            this._io.seek(_pos);
            return this.dummy;
        }

As I don't need any parsing, I thought of misusing value instances and that doesn't work. The following KSY...

      dummy2:
        value: fmt_oms_apl_data_date_g_meh(_root, fmt_oms_dll_device_type::heat_meter_outlet.to_i, fmt_oms_dll_device_type::water_meter.to_i)

...results in the following exception:

	Exception in thread "main" scala.MatchError: Name(identifier(fmt_oms_apl_data_date_g_meh)) (of class io.kaitai.struct.exprlang.Ast$expr$Name)
		at io.kaitai.struct.translators.TypeDetector.detectTypeRaw(TypeDetector.scala:165)
		at io.kaitai.struct.translators.TypeDetector.detectType(TypeDetector.scala:25)
		at io.kaitai.struct.precompile.ValueTypesDeriver.$anonfun$deriveValueType$2(ValueTypesDeriver.scala:27)
		at io.kaitai.struct.precompile.ValueTypesDeriver.$anonfun$deriveValueType$2$adapted(ValueTypesDeriver.scala:20)
		at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:231)
		at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
		at io.kaitai.struct.precompile.ValueTypesDeriver.deriveValueType(ValueTypesDeriver.scala:20)
		at io.kaitai.struct.precompile.ValueTypesDeriver.$anonfun$deriveValueType$5(ValueTypesDeriver.scala:50)
		at io.kaitai.struct.precompile.ValueTypesDeriver.$anonfun$deriveValueType$5$adapted(ValueTypesDeriver.scala:48)
		at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:231)
		at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
		at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
		at io.kaitai.struct.precompile.ValueTypesDeriver.deriveValueType(ValueTypesDeriver.scala:48)
		at io.kaitai.struct.precompile.ValueTypesDeriver.run(ValueTypesDeriver.scala:12)
		at io.kaitai.struct.precompile.SpecsValueTypeDerive.$anonfun$run$2(SpecsValueTypeDerive.scala:15)
		at io.kaitai.struct.precompile.SpecsValueTypeDerive.$anonfun$run$2$adapted(SpecsValueTypeDerive.scala:13)
		at scala.collection.mutable.HashMap.$anonfun$foreach$1(HashMap.scala:138)
		at scala.collection.mutable.HashTable.foreachEntry(HashTable.scala:236)
		at scala.collection.mutable.HashTable.foreachEntry$(HashTable.scala:229)
		at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:40)
		at scala.collection.mutable.HashMap.foreach(HashMap.scala:138)
		at io.kaitai.struct.precompile.SpecsValueTypeDerive.run(SpecsValueTypeDerive.scala:13)
		at io.kaitai.struct.Main$.precompile(Main.scala:38)
		at io.kaitai.struct.Main$.$anonfun$importAndPrecompile$3(Main.scala:28)
		at io.kaitai.struct.Main$.$anonfun$importAndPrecompile$3$adapted(Main.scala:27)
		at scala.collection.mutable.HashMap.$anonfun$foreach$1(HashMap.scala:138)
		at scala.collection.mutable.HashTable.foreachEntry(HashTable.scala:236)
		at scala.collection.mutable.HashTable.foreachEntry$(HashTable.scala:229)
		at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:40)
		at scala.collection.mutable.HashMap.foreach(HashMap.scala:138)
		at io.kaitai.struct.Main$.$anonfun$importAndPrecompile$1(Main.scala:27)
		at io.kaitai.struct.Main$.$anonfun$importAndPrecompile$1$adapted(Main.scala:24)
		at scala.util.Success.$anonfun$map$1(Try.scala:251)
		at scala.util.Success.map(Try.scala:209)
		at scala.concurrent.Future.$anonfun$map$1(Future.scala:289)
		at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:29)
		at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:29)
		at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
		at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:140)
		at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
		at java.util.concurrent.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1021)
		at java.util.concurrent.ForkJoinPool$WorkQueue.execLocalTasks(ForkJoinPool.java:1046)
		at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1058)
		at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
		at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)

Am I doing something wrong with the `value`- instance? If not, do you think my use case is worth supporting? It looks like a classic `value`- instance to me in theory.

In my opinion my use case might even be related to a question I asked in the past already about designs of CTORs of types regarding implicit `_io`, `_root` and stuff compared to explicit `params`:

https://github.com/kaitai-io/kaitai_struct/issues/275#issuecomment-368820503

Using `params`  only, one could get rid of the `_io`-stuff in the checking types and not even parse by accident anymore, because there wouldn't be any data to parse from given using the CTORs.

What do you think?

**Comment by webbnh** [comment]
While we wait for an authoritative answer, let me pose a small question.

Given
```
instances:
  dummy:
    pos:  _io.pos
    type: fmt_oms_apl_data_date_g_meh(_root, fmt_oms_dll_device_type::heat_meter_outlet.to_i, fmt_oms_dll_device_type::water_meter.to_i)

  dummy2:
    value: fmt_oms_apl_data_date_g_meh(_root, fmt_oms_dll_device_type::heat_meter_outlet.to_i, fmt_oms_dll_device_type::water_meter.to_i)
```
isn't `dummy2` trying to return a _type_ where it should be returning a _value_?

**Comment by ams-tschoening** [comment]
Yes, but what's the difference between both from my point of view? If value would be the constant `1`, KS would happily store that value with the type `Integer`. `value` could be some string as well or some logical expression etc. In the end, it would simply be the result of some calculation assigned to some type.

So why not let that result be the instanciation of some (KS-)type only as well? The only difference between my examples and the KS-types is that KS-types are primarily designed to parse things. But that might be an unnecessary restriction and I'm even able to "work around" that using `size: 0` and stuff. So I'm already able to create KS-types not parsing anything and misuse them instead using `type`-instances. :-) It only breaks for `value`-instances because of limitations of the translator.

Look at the generated Java code for the `dummy`-instance I posted above and compare that to some arbitrary `value`-instance:

        private Boolean isDateGBase;
        public Boolean isDateGBase() {
            if (this.isDateGBase != null)
                return this.isDateGBase;
            boolean _tmp = (boolean) ( ((vdbHeader().dib().dif().dataCoding() == FmtRecOms.FmtOmsAplDibDataCoding.BIN_INT_16_BITS) && (vdbHeader().vib().unit() == FmtRecOms.FmtOmsAplVibUnit.DATE_G)) );
            this.isDateGBase = _tmp;
            return this.isDateGBase;
        }

It's pretty much the same, only with the seeking overhead. Both return some type in the end, only different ones. That's why I asked myself if it wouldn't be a feature to allow `value`-instances like in my example. The way I'm using the type in my `value`-instance even works for `case` already:

      - id:           data
        size-eos:     true
        type:
          switch-on:  data_type
          cases:
            fmt_oms_apl_data_type::date_g_meh: fmt_oms_apl_data_date_g_meh(_root, fmt_oms_dll_device_type::heat_meter_outlet.to_i, fmt_oms_dll_device_type::water_meter.to_i)

So, there's already some support for arbitrary type instanciation available and the question is if it should be enhanced to support `value`-instances as well.

**Comment by GreyCat** [comment]
From what I see, you're basically trying to:

1. Invoke a construction of object, passing a given set of arguments to constructor
2. Use Python-like function call syntax for that — i.e. `ClassName(arg1, arg2, ...)`

KS expression language uses indeed Python-like syntax, so your expression passes syntax analysis stage, and then it is basically represented as function call — i.e.:

```
fmt_oms_apl_data_date_g_meh(_root, arg2, arg3)
```

The obvious problem is that we don't support *any* function calls right now, and, even if we will be supporting something like that, we'll be likely introducing some extra syntax to differentiate between a function call vs invocation of a constructor.

So, to summarize it:

* Creation of new objects in expressions is not yet implemented, and I'm not even sure if we'll ever be implementing that (for example, in C, C++ or Rust these will raise serious questions of ownership of these objects and relevant memory management)
* If we will be doing that, there will be likely some special syntax to do that.

**Comment by ams-tschoening** [comment]
> * Creation of new objects in expressions is not yet implemented, and I'm not even sure if we'll ever be implementing that (for example, in C, C++ or Rust these will raise serious questions of ownership of these objects and relevant memory management)

How do you answer the questions of memory management for KS-types and parsing `type`-instances in those languages currently? I don't see any difference to what I'm asking about and Java generates pretty much the same code, storing the parsing type in the instance.

> * If we will be doing that, there will be likely some special syntax to do that.

Do you really think that's necessary? The current syntax from my example would be in line with all other instanciations of KS-types. The only difference is that some types parse content and some don't in my opinion and that could even be hidden as an implementation details depending on if `_io` is forwarded or not.

**Comment by GreyCat** [comment]
> How do you answer the questions of memory management for KS-types and parsing `type`-instances in those languages currently?

"Parse instances", i.e. those having type are owned (and stored) by the object they belong to. They are specifically constructed in that invocation, and then stored as a member of the object. They are destructed when their owning object is being destroyed.

"Value instances", on the other hand, just provide non-owning references to any objects, so they should not create anything new. Consider the following example:

```yaml
instances:
  a:
    pos: 0x100
    type: foo
  b:
    value: a
```

In this example, only instance `a` is constructed, stored and owned inside this type's object. `b` instance returns the reference, so it's never stored anywhere (and never destroyed) — we just reference another existing object.

> Do you really think that's necessary? The current syntax from my example would be in line with all other instanciations of KS-types.

`type` and `value` are totally different things and they're likely to stay that way. `value` specifies a KS expression language text. `type` is parsed specially, that field only allows references to existing type names (which are resolved in early precompile time) and an optional comma-separated list of KS expressions to be used as parameters during creation of new instance.

**Comment by ams-tschoening** [comment]
> "Value instances", on the other hand, just provide non-owning references to any objects, so they should not create anything new. Consider the following example:[...]

But they already support creating entirely new values, just have a look at the example I brought into discussion [some comments above](https://github.com/kaitai-io/kaitai_struct/issues/521#issuecomment-462402867) already:

    private Boolean isDateGBase;
    public Boolean isDateGBase() {
        if (this.isDateGBase != null)
            return this.isDateGBase;
        boolean _tmp = (boolean) ( ((vdbHeader().dib().dif().dataCoding() == FmtRecOms.FmtOmsAplDibDataCoding.BIN_INT_16_BITS) && (vdbHeader().vib().unit() == FmtRecOms.FmtOmsAplVibUnit.DATE_G)) );
        this.isDateGBase = _tmp;
        return this.isDateGBase;
    }

The `value`-instance is creating a completely new value for `isDateGBase`, which didn't exist before and is therefore coupled to the lifetime of the owning parent type. That's exactly what is done in case of `parse`-instances as well, only different types are created and stored. All other concepts are exactly the same, even the generated code is almost the same already. Which makes sense, because pretty much the same `instance*`-methods of the language compiler are used to render the code in all cases.

**Comment by GreyCat** [comment]
`isDateGBase` is a boolean, not a user type. Creation of booleans, integers, floats, strings is treated as primitive types, so there's really no need to introduce the concept of ownership here. They're effectively immutable and can be copied easily.

User types, however, need to be owned by some object, as they need to be properly deleted. So, the only way to create user type objects we offer is via seq attribute / parse instance, this way we always know who's "owning" this object.

You can actually use parse instances right now without any `pos` or `size`. That would be effectively what you want:

```yaml
instances:
  dummy:
    type: fmt_oms_apl_data_date_g_meh(_root, fmt_oms_dll_device_type::heat_meter_outlet.to_i, fmt_oms_dll_device_type::water_meter.to_i)
```

If `fmt_oms_apl_data_date_g_meh` is not doing any `seq` parsing, that's absolutely fine, we just use it to keep track of ownership.

**Comment by KOLANICH** [comment]
>Creation of booleans, integers, floats, strings is treated as primitive types, so there's really no need to introduce the concept of ownership here.

But they are still owned:

```cpp
struct A{
double a;
};
```
`a` still consumes memory.

```cpp
struct A{
std::string a;
};

Is owned too, destructed automatically.

So what's wrong with owning? C++ does destructing itself, when RAII is used.

**Comment by ams-tschoening** [comment]
> You can actually use parse instances right now without any `pos` or `size`. That would be effectively what you want:

I know, `pos` in my example is only used to guarantee no influence on parsing and the stream of data even by accident. But that's the point of my discussion: Using a type more or less designed to parse could be considered a misuse, while a type not able to parse at all would perfectly fit my use case. Using that, any attempt to parse e.g. using `seq` would result in some error as well.

So I thought of distinguishing between parsing and non-parsing types as concepts, because you have everything in place in theory to support both: The lifetime and memory management of both is actually the same, parsing vs. non-parsing could in theory be distinguished using another keyword or the context or even using the [concept of params](https://github.com/kaitai-io/kaitai_struct/issues/275#issuecomment-368820503).

`value` was simply one example for distinguishing by context, because in that case the KSY-author already made clear that one is not interested in any parsing and KS could create a non-parsing type in that special case. Using e.g. `value` it's even directly documented that no parsing happens there, while with only using `type` one can't be sure until looking at the mentioned type.

Examples for additional keywords could be `type-non-reading|parsing` or such. `params` OTOH could be used to tell KS which CTORs to generate for types at all and for non-parsing types one might simply not add `_io`, so there can't be any parsing even by accident.

In the end, it's more about what is considered a hack/workaround/misuse or supported by KS and documentation stuff and preventing accidents and all that.

**Comment by webbnh** [comment]
I apologize for not being able to engage properly with this discussion, and so I'm sure that I've missed the important point, but, @ams-tschoening, wouldn't it suffice for your purposes if your value instance contained an expression which returned a (e.g., boolean) _value_?

I understand that, in all right-thinking languages, instances of user types are themselves values (and, even, types themselves are values), but, since KS tries to cater to something approaching _all_ languages, it seems like having your expression evaluate to a value with a type which is intrinsic to the supported languages would give you the result that you want while affording KS the freedom that it needs.

**Comment by ams-tschoening** [comment]
> [...]wouldn't it suffice for your purposes if your value instance contained an expression which returned a (e.g., boolean) _value_?

You mean something like the following?

    instances:
        is_affected_by_bug_xy:
            value: bug_checker(_root, arg1, arg2).is_affected_by_xy

Instead of:

    instances:
        get_bug_checker:
             value: bug_checker(_root, arg1, arg2)

And accessing `get_bug_checker.is_affected_by_xy` somewhere?

I'm pretty sure the first thing doesn't work as well, but didn't try it. I focussed on the second example because it's closer to what already exists, one doesn't need to instantiate `bug_checker` more than once and simply because one doesn't need to create all the `is_affected_by_bug_xy` instances.

> [...]but, since KS tries to cater to something approaching _all_ languages[...]

The question is if KS already supports everything needed for what I have in mind and in my opinion it has, because I'm very close to parsing instances using parsing types. So I don't see any incompatibilities with target languages or such, it's only a design decision if KS wants to support it or not. As @GreyCat already mentioned, it's somewhat possible already, only in less good documented and a bit more error-prone in theory.
